hal/src/thumbv7em/clock/v2/xosc.rs

```diff
@@ -311,37 +267,83 @@ impl<X: XoscNum> XoscToken<X> {
 pub type StartUp = STARTUP_A;
 
 /// Type alias for Xosc Input pin
-pub type XIn<X> = Pin<<X as XoscNum>::XIn, FloatingDisabled>;
+pub type XIn<X> = Pin<<X as XoscId>::XIn, FloatingDisabled>;
 
 /// Type alias for Xosc Output pin
-pub type XOut<X> = Pin<<X as XoscNum>::XOut, FloatingDisabled>;
+pub type XOut<X> = Pin<<X as XoscId>::XOut, FloatingDisabled>;
 
 //==============================================================================
 // Mode structure for Xosc
 //==============================================================================
 
-/// Trait that defines a mode [`Xosc`] is operating in
-pub trait Mode: Sealed {}
+mod private {
+    use super::{CrystalCurrent, Sealed, XOut, XoscId, XoscToken};
 
-/// Struct representing a clock mode for [`Xosc`]
+    /// Private trait to restrict access to the `configure` function
+    pub trait Mode<X: XoscId>: Sealed {
+        const FROM_CRYSTAL: bool;
```

Comment:
Well, I wanted to preserve the order of calls within `.enable()` while also keeping as little redundant code as necessary. There might be a way to restructure this.

In general though, what do you think about the approach [here](https://github.com/atsamd-rs/atsamd/blob/16ff6e0afd3e5e01084149ad033ac04f5eb9fd82/hal/src/sercom/v2/spi.rs#L475-L488)? Each implementer of `Mode` maps to the corresponding PAC constants. To me, that makes perfect sense. You're mapping from types to values.

Would it help if the `bool` were named `XTALEN`, which matches the PAC field name? That way it's more clear that the type is mapping to the corresponding register fields.

Reply:
That'd work better IMO because associated constant provides actual data, not the condition value that is used to execute one code branch or the other somewhere else. However, it does not feel right that mode specfic configuration is partially done on a `Xosc::enable` level by using data stored in associated constants and partially in `fn configure/enable` level. It feels inconsistent. I like `mode` specific stuff to be hidden away in `fn configure/enable` although I admit you're absolutely right that order of execution matters which in that design is pure abstraction leakage and implementor must magically know to put `self.mode.enable/configure(&mut self.token)` in a proper spot in `Xosc::enable`. Other approach would assume that you get rid of `fn configure` altogether but then you end up with code repetition and nasty ifs. Hmm, tricky.

Reply:
I understand your objection better now. And I agree. It probably is an anti-pattern to use an associated `bool` in a condition.

I think my goal will be to change the implementation so that `Mode<X>` *only* abstracts data and does not abstract behavior. I think that makes for the cleanest design. I think I actually performed a similar refactor in the `spi` module long ago. I used to have `configure` functions on the `Mode` trait, but I removed them in favor of the data-only approach.

Thanks for helping to clarify that.

Reply:
I think the latest implementation resolves this issue.